### PR TITLE
Add streaming graph persistence layer from kafka to file

### PR DESCRIPTION
**Pull request by tmkasun**
In this PR:

- Add `JasmineGraphInstanceProtocol` command to sent streaming graph edges from `frontendservicesesion` to worker's `instanceservicesession`
image

### CODE REVIEW

File: src/centralstore/incremental/DataPublisher.cpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    this->worker_address = worker_address;
+    serv_addr.sin_family = AF_INET;
+    serv_addr.sin_port = htons(worker_port);
+    if (inet_pton(AF_INET, worker_address.c_str(), &serv_addr.sin_addr) <= 0) {
+        data_publisher_logger.error("Invalid address/ Address not supported!");
+    }
+}
+
+void DataPublisher::publish(std::string message) {
+    char recever_buffer[1024] = {0};
```
Comment (miyurud) [inline src/centralstore/incremental/DataPublisher.cpp:19]: Lets declare the 1024 as a constant. See for [example](https://github.com/miyurud/jasminegraph/blob/master/src/frontend/JasmineGraphFrontEndProtocol.h#L85).

File: src/centralstore/incremental/DataPublisher.cpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    // Sending edge data content length
+    send(this->sock, &converted_number, sizeof(converted_number), 0);
+
+    int received_int = 0;
+    data_publisher_logger.info("Waiting for content length ack\n");
+    auto return_status = recv(this->sock, &received_int, sizeof(received_int), 0);
+    // Receve ack for edge data content length
+
+    if (return_status > 0) {
+        fprintf(stdout, "Received int = %d\n", ntohl(received_int));
```
Comment (miyurud) [inline src/centralstore/incremental/DataPublisher.cpp:46]: I do not see any reason for having this `fprintf()` here. The message is also not intuitive.

File: src/centralstore/incremental/DataPublisher.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
+#include <arpa/inet.h>
```
Comment (miyurud) [inline src/centralstore/incremental/DataPublisher.h:1]: Add copyright header

File: src/centralstore/incremental/NodeBlock.cpp [hunk 4]
```diff
@@ -221,6 +223,11 @@ NodeBlock* NodeBlock::get(unsigned int blockAddress) {
         node_block_logger.error("Error while reading edge reference data from block " + std::to_string(blockAddress));
```
Comment (miyurud) [inline src/centralstore/incremental/NodeBlock.cpp:223]: Don't we have to say which graph ID and which partition ID we are referring to here?

File: src/centralstore/incremental/NodeBlock.cpp [hunk 5]
```diff
@@ -221,6 +223,11 @@ NodeBlock* NodeBlock::get(unsigned int blockAddress) {
...
     }
 
+    if (!NodeBlock::nodesDB->read(reinterpret_cast<char*>(&edgeRefPID), sizeof(unsigned char))) {
+        node_block_logger.error("Error while reading edge reference partition ID data from block " +
```
Comment (miyurud) [inline src/centralstore/incremental/NodeBlock.cpp:227]: Don't we have to say which graph ID and which partition ID we are referring to here?

File: src/frontend/JasmineGraphFrontEnd.cpp [hunk 6]
```diff
@@ -17,6 +17,7 @@ limitations under the License.
 #include <map>
 #include <set>
 #include "JasmineGraphFrontEnd.h"
+#include <nlohmann/json.hpp>
```
Comment (miyurud) [inline src/frontend/JasmineGraphFrontEnd.cpp:20]: This is a new library for JasmineGraph. We need to know the license and also update the information in the README.

File: src/frontend/JasmineGraphFrontEnd.cpp [hunk 7]
```diff
@@ -50,6 +53,7 @@ void *frontendservicesesion(std::string masterIP, int connFd, SQLiteDBInterface
     frontend_logger.log("Master IP: " + masterIP, "info");
     char data[FRONTEND_DATA_LENGTH];
     bzero(data, FRONTEND_DATA_LENGTH + 1);
+    DataPublisher workerClient(7780, "127.0.0.1");
```
Comment (miyurud) [inline src/frontend/JasmineGraphFrontEnd.cpp:56]: Why is this host IP and port hard coded here?

File: src/frontend/JasmineGraphFrontEnd.cpp [hunk 8]
```diff
@@ -515,7 +518,17 @@ void *frontendservicesesion(std::string masterIP, int connFd, SQLiteDBInterface
...
                         "info");
-                graphPartitioner.addEdge(edge);
+                auto edgeJson = json::parse(data);
+                auto sourceJson = edgeJson["source"];
+                auto destinationJson = edgeJson["destination"];
+
+                std::string sId = std::string(sourceJson["id"]);
+                std::string dId = std::string(destinationJson["id"]);
+
+                partitionedEdge pe = graphPartitioner.addEdge({sId, dId});
```
Comment (miyurud) [inline src/frontend/JasmineGraphFrontEnd.cpp:528]: Lets have better variable name than `pe`

File: src/localstore/incremental/JasmineGraphIncrementalLocalStore.cpp [hunk 9]
```diff
@@ -0,0 +1,29 @@
+#include <memory>
```
Comment (miyurud) [inline src/localstore/incremental/JasmineGraphIncrementalLocalStore.cpp:1]: Missing license header

File: src/localstore/incremental/JasmineGraphIncrementalLocalStore.cpp [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    gc.openMode = "trunk";
+    this->nm = new NodeManager(gc);
+};
+
+std::string JasmineGraphIncrementalLocalStore::addEdgeFromString(std::string edgeString)
+{
+    try {
+        auto edge = json::parse(edgeString);
+        NodeBlock *sourceNode = this->nm->addNode(edge["id"]);
+        // NodeBlock* destinationNode = nodeManagers.at(destinationJson["pid"].get<int>())->addNode(dId);
```
Comment (miyurud) [inline src/localstore/incremental/JasmineGraphIncrementalLocalStore.cpp:24]: Remove the commented code

File: src/localstore/incremental/JasmineGraphIncrementalLocalStore.cpp [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    } catch (const std::exception &) {
+        incremental_localstore_logger.log("Error malformed JSON attributes!", "error");
+        // TODO handle JSON errors
```
Comment (miyurud) [inline src/localstore/incremental/JasmineGraphIncrementalLocalStore.cpp:27]: When you type a TODO please use the format ``TODO<github-user-name>: message``

File: src/localstore/incremental/JasmineGraphIncrementalLocalStore.h [hunk 12]
```diff
@@ -0,0 +1,16 @@
+#include <string>
```
Comment (miyurud) [inline src/localstore/incremental/JasmineGraphIncrementalLocalStore.h:1]: Add license header

File: src/partitioner/stream/Partition.cpp [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+        for (auto edgeCuts : partition) {
+            total += edgeCuts.second.size();
+        }
+    }
+    return total;
+}
+
+float Partition::edgeCutsRatio() { return this->edgeCutsCount() / (this->getEdgesCount() + this->edgeCutsCount()); }
+
+void Partition::printEdgeCuts() {
```
Comment (miyurud) [inline src/partitioner/stream/Partition.cpp:112]: This function seems to be for debugging purpose. Shall we make the `cout` statements to debug log lines?

File: src/partitioner/stream/Partition.cpp [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+            std::cout << edgeList.first << " ____" << std::endl;
+            for (std::string vertext : edgeList.second) {
+                std::cout << "\t| ---> " << vertext << std::endl;
+            }
+            std::cout << "\n" << std::endl;
+        }
+    }
+}
+
+void Partition::printEdges() {
```
Comment (miyurud) [inline src/partitioner/stream/Partition.cpp:125]: The same comment on `cout` applies here too.

File: src/partitioner/stream/Partitioner.cpp [hunk 15]
```diff
@@ -209,8 +210,8 @@ partitionedEdge Partitioner::fennelPartitioning(std::pair<int, int> edge) {
  * Expect a space seperated pair of vertexts representing an edge in the graph.
  **/
 std::pair<long, long> Partitioner::deserialize(std::string data) {
-    std::vector<std::string> v = JasmineGraphIncrementalStore::_split(data, ' ');
-    // std::cout << "Vertext 1 = " << stoi(v[0]) << std::endl;
-    // std::cout << "Vertext 2 = " << stoi(v[1]) << std::endl;
+    std::vector<std::string> v = Partition::_split(data, ' ');
+    std::cout << "Vertext 1 = " << stoi(v[0]) << std::endl;
```
Comment (miyurud) [inline src/partitioner/stream/Partitioner.cpp:214]: Why do we need these couts?

**Review by miyurud — CHANGES_REQUESTED** [review]

File: src/centralstore/incremental/NodeBlock.cpp (hunk 4 above)
Comment (tmkasun) [reply src/centralstore/incremental/NodeBlock.cpp:223]: I wish we could, But in this class, we don't have graph & partition information, Graph ID and Partition ID is used to create the NodeDB database file so it's in the file name, But in this class, all we have is the file stream [`std::fstream*`](https://github.com/miyurud/jasminegraph/blob/740e27dd015904c4fd14483546dcedafc25686e9/src/centralstore/incremental/NodeBlock.cpp#L264), to the DB file , But it seems we can't get the file path from `fstream` or at least it's not straightforward

- [SO Answer 1](https://stackoverflow.com/questions/3022692/how-get-file-path-from-stdifstream-in-c)
- [SO Answer 2](https://stackoverflow.com/questions/10773391/getting-filename-or-path-from-fstream)

File: src/centralstore/incremental/NodeBlock.cpp (hunk 5 above)
Comment (tmkasun) [reply src/centralstore/incremental/NodeBlock.cpp:227]: Not at this NodeDB class level as mentioned in https://github.com/miyurud/jasminegraph/pull/122#discussion_r606796318

File: src/frontend/JasmineGraphFrontEnd.cpp (hunk 6 above)
Comment (tmkasun) [reply src/frontend/JasmineGraphFrontEnd.cpp:20]: This has [MIT license](https://github.com/nlohmann/json/blob/develop/LICENSE.MIT), Will update the README file as well

File: src/partitioner/stream/Partitioner.cpp (hunk 15 above)
Comment (tmkasun) [reply src/partitioner/stream/Partitioner.cpp]: These are for debugging purpose, Hence converted them to debug logs
Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/bdd734e88390169ddc9e44c74030d1c1b170d0ae

File: src/partitioner/stream/Partition.cpp (hunk 14 above)
Comment (tmkasun) [reply src/partitioner/stream/Partition.cpp:132]: Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/bdd734e88390169ddc9e44c74030d1c1b170d0ae

File: src/partitioner/stream/Partition.cpp (hunk 13 above)
Comment (tmkasun) [reply src/partitioner/stream/Partition.cpp:115]: Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/bdd734e88390169ddc9e44c74030d1c1b170d0ae

File: src/localstore/incremental/JasmineGraphIncrementalLocalStore.h (hunk 12 above)
Comment (tmkasun) [reply src/localstore/incremental/JasmineGraphIncrementalLocalStore.h:14]: Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/bdd734e88390169ddc9e44c74030d1c1b170d0ae

File: src/localstore/incremental/JasmineGraphIncrementalLocalStore.cpp (hunk 11 above)
Comment (tmkasun) [reply src/localstore/incremental/JasmineGraphIncrementalLocalStore.cpp]: Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/bdd734e88390169ddc9e44c74030d1c1b170d0ae

File: src/centralstore/incremental/DataPublisher.h (hunk 3 above)
Comment (tmkasun) [reply src/centralstore/incremental/DataPublisher.h:14]: Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/bdd734e88390169ddc9e44c74030d1c1b170d0ae

File: src/centralstore/incremental/DataPublisher.cpp (hunk 1 above)
Comment (tmkasun) [reply src/centralstore/incremental/DataPublisher.cpp]: Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/bdd734e88390169ddc9e44c74030d1c1b170d0ae

File: src/centralstore/incremental/DataPublisher.cpp [hunk 16]
```diff
@@ -0,0 +1,53 @@
+#include "./DataPublisher.h"
```
[the remark this answers is not in the document]
Comment (tmkasun) [reply src/centralstore/incremental/DataPublisher.cpp:14]: Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/bdd734e88390169ddc9e44c74030d1c1b170d0ae

File: src/frontend/JasmineGraphFrontEnd.cpp (hunk 6 above)
Comment (tmkasun) [reply src/frontend/JasmineGraphFrontEnd.cpp:20]: Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/bdd734e88390169ddc9e44c74030d1c1b170d0ae

File: src/centralstore/incremental/DataPublisher.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (miyurud) [reply src/centralstore/incremental/DataPublisher.cpp:14]: Done

File: src/centralstore/incremental/DataPublisher.cpp (hunk 1 above)
Comment (miyurud) [reply src/centralstore/incremental/DataPublisher.cpp]: Done

File: src/centralstore/incremental/DataPublisher.h (hunk 3 above)
Comment (miyurud) [reply src/centralstore/incremental/DataPublisher.h:14]: Done

File: src/frontend/JasmineGraphFrontEnd.cpp (hunk 6 above)
Comment (miyurud) [reply src/frontend/JasmineGraphFrontEnd.cpp:20]: Ok

File: src/frontend/JasmineGraphFrontEnd.cpp (hunk 8 above)
Comment (miyurud) [reply src/frontend/JasmineGraphFrontEnd.cpp]: Done

File: src/localstore/incremental/JasmineGraphIncrementalLocalStore.cpp (hunk 10 above)
Comment (miyurud) [reply src/localstore/incremental/JasmineGraphIncrementalLocalStore.cpp]: Done

File: src/localstore/incremental/JasmineGraphIncrementalLocalStore.cpp (hunk 9 above)
Comment (miyurud) [reply src/localstore/incremental/JasmineGraphIncrementalLocalStore.cpp:14]: Done

File: src/localstore/incremental/JasmineGraphIncrementalLocalStore.cpp (hunk 11 above)
Comment (miyurud) [reply src/localstore/incremental/JasmineGraphIncrementalLocalStore.cpp]: Done

File: src/localstore/incremental/JasmineGraphIncrementalLocalStore.h (hunk 12 above)
Comment (miyurud) [reply src/localstore/incremental/JasmineGraphIncrementalLocalStore.h:14]: Done

File: src/partitioner/stream/Partition.cpp (hunk 13 above)
Comment (miyurud) [reply src/partitioner/stream/Partition.cpp:115]: Done

File: src/partitioner/stream/Partition.cpp (hunk 14 above)
Comment (miyurud) [reply src/partitioner/stream/Partition.cpp:132]: Done

File: src/partitioner/stream/Partitioner.cpp (hunk 15 above)
Comment (miyurud) [reply src/partitioner/stream/Partitioner.cpp]: Ok, Done

File: src/centralstore/incremental/DataPublisher.cpp (hunk 2 above)
Comment (tmkasun) [reply src/centralstore/incremental/DataPublisher.cpp]: Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/bdd734e88390169ddc9e44c74030d1c1b170d0ae

**Comment by tmkasun** [comment]
> The use of the variable `JASMINE_HOME` here is ambiguous. Shouldn't this be just user's home directory?

As per the offline discussion , Renamed the `JASMINE_HOME` environment variable name to `JASMINEGRAPH_HOME` to prevent conflicting with default user `HOME` environment variable

Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/72bf122c4e928dfcf8739b15f2c51f67012a5c21

File: src/centralstore/incremental/NodeBlock.cpp (hunk 4 above)
Comment (miyurud) [reply src/centralstore/incremental/NodeBlock.cpp:223]: Based on the discussion we had today, it seems that it is not possible to link the graph ID and partition ID from the execution context. Hence we will keep this error message as it is and proceed.

File: src/centralstore/incremental/NodeBlock.cpp (hunk 5 above)
Comment (miyurud) [reply src/centralstore/incremental/NodeBlock.cpp:227]: Ok

File: src/frontend/JasmineGraphFrontEnd.cpp (hunk 7 above)
Comment (tmkasun) [reply src/frontend/JasmineGraphFrontEnd.cpp]: Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/7174d9c7814f01e19de59293f8917ecd669a662f

Comment (miyurud) [reply src/frontend/JasmineGraphFrontEnd.cpp]: Ok, done

**Comment by miyurud** [comment]
> > The use of the variable `JASMINE_HOME` here is ambiguous. Shouldn't this be just user's home directory?
> 
> As per the offline discussion , Renamed the `JASMINE_HOME` environment variable name to `JASMINEGRAPH_HOME` to prevent conflicting with default user `HOME` environment variable
> 
> Fixed in [72bf122](https://github.com/miyurud/jasminegraph/commit/72bf122c4e928dfcf8739b15f2c51f67012a5c21)

Ok, but we should keep `HOME` variable as it is. See the comment I have put there.

**Review by miyurud — APPROVED** [review]

**Pull request merged**